openstreetmap-carto.lua

```diff
@@ -0,0 +1,518 @@
+    if filter == 1 then
+        return 1, keyvalues
+    end
+
+    -- Filter out all relations except route, multipolygon and boundary relations
+    if ((keyvalues["type"] ~= "route") and (keyvalues["type"] ~= "multipolygon") and (keyvalues["type"] ~= "boundary")) then
+        return 1, keyvalues
+    end
+
+    return 0, keyvalues
+end
+
+-- Filtering on ways
+function filter_tags_way (keyvalues, numberofkeys)
+    local filter = 0  -- Will object be filtered out?
+    local polygon = 0 -- Will object be treated as polygon?
+
+    -- Filter out objects that are filtered out by filter_tags_generic
+    filter, keyvalues = filter_tags_generic(keyvalues, numberofkeys)
+    if filter == 1 then
+        return filter, keyvalues, polygon, roads
+    end
+
+    polygon = isarea(keyvalues)
+
+    -- Add z_order column
+    keyvalues.z_order = z_order(keyvalues)
+
+    return filter, keyvalues, polygon, roads(keyvalues)
+end
+
+--- Handling for relation members and multipolygon generation
+-- Multipolygons are either old-style or new-style. As defined here,
+-- a new-style MP is one with any tags other than type on the relation, and for
+-- them the tags of the ways do not matter.
+--
+-- An old-style MP is one where the only tag on the relation is
+-- type=multipolygon and the tags of all the members are either the same or
+-- empty.
+--
+-- These are stricter definitions then have been used in the past by the C
+-- transforms, but cut down on bugs where a new-style MP suddenly gets treated
+-- as an old-style MP.
+--
+-- This has a few properties
+--
+-- - MP generation does not depend on polygon characteristics of tags on ways
+--
+-- - All tags are considered, except for deleted tags which are ignored for
+--   technical reasons
+--
+-- - Any new-style MP will never change to an old-style MP by only changing
+--   tags on its member ways, nor will it become invalid
+--
+-- - Any old-style MP will never change to an new-style MP by only changing
+--   tags on its member ways, but may become invalid if the tags become
+--   conflicting
+--
+-- - All multipolygons are polygons
+--
+-- @param keyvalues OSM tags, after processing by relation transform
+-- @param keyvaluemembers OSM tags of relation members, after processing by way transform
+-- @param roles OSM roles of relation members
+-- @param membercount number of members
+-- @return filter, cols, member_superseded, boundary, polygon, roads
+function filter_tags_relation_member (keyvalues, keyvaluemembers, roles, membercount)
+    local members_superseeded = {}
+
+    -- Start by assuming that this not an old-style MP
+    for i = 1, membercount do
+        members_superseeded[i] = 0
+    end
+
+    local type = keyvalues["type"]
+
+    -- Remove type key
+    keyvalues["type"] = nil
+
+    -- Boundary relations are treated as linestring
+    if type == "boundary" or (type == "multipolygon" and keyvalues["boundary"]) then
```

Comment:
This is from https://github.com/openstreetmap/osm2pgsql/blob/master/tagtransform.cpp#L207-L209, which regards `type=multipolygon boundary=*` the same as `type=boundary` for osm2pgsql geometry formation (forming linestrings from polygons), but not `type=multipolygon` and `type=boundary` without a `boundary=*` tag on either.

Does this make sense? Maybe.